File: Programmers/[67260] 동굴 탐험/KingDonggyu/67260.js
function solution(n, path, order) {
  const graph = new Array(n).fill(null).map((_) => []);

  path.forEach(([x, y]) => {
    graph[x].push(y);
    graph[y].push(x);
  });

  const beforeRoom = new Array(n).fill(null); // 먼저 방문해야 하는 방
  const afterRoom = new Array(n).fill(null); // 다음 방문할 수 있는 방

  order.forEach(([x, y]) => {
    beforeRoom[y] = x;
    afterRoom[x] = y;
  });

  // 0보다 먼저 방문해야 하는 방이 있는지 확인
  if (beforeRoom[0]) {
    return false;
  }

  const visited = new Array(n).fill(false);
  const isWaiting = new Array(n).fill(false);
  const stack = [0];
  let roomCount = 1;
  visited[0] = true;

  const visitRoom = (x) => {
    stack.push(x);
    visited[x] = true;
    roomCount++;
  };

  // DFS
  while (stack.length > 0) {
    const x = stack.pop();

    for (const next of graph[x]) {
      if (visited[next]) {
        continue;
      }

      if (beforeRoom[next] && !visited[beforeRoom[next]]) {
        // 다른 방을 먼저 탐색하도록 wait
        isWaiting[next] = true;
        continue;
      }

      visitRoom(next);

      const after = afterRoom[next];
      if (after && isWaiting[after]) {
        // wait 중인 방을 활성화
        visitRoom(after);
        isWaiting[after] = false;
      }
    }
  }

  return roomCount === n;
}

File: Programmers/[67260] 동굴 탐험/SUbbb/67260.java
import java.util.*;

class Solution {
    boolean[] visited;
    int[] before;
    int[] after;
    ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();
    
    public boolean solution(int n, int[][] path, int[][] order) {
        visited = new boolean[n];
        before = new int[n];
        after = new int[n];
        
        // 방문 우선순위 저장
        for (int[] or : order) before[or[1]] = or[0];
        
        for (int i = 0; i < n; i++) adjList.add(new ArrayList<>());
        
        // 인접 리스트 생성
        for (int[] p : path) {
            adjList.get(p[0]).add(p[1]);
            adjList.get(p[1]).add(p[0]);
        }
        
        // 0보다 먼저 방문해야 하는 방이 있다면 바로 false
        if (before[0] != 0) return false;
        
        dfs();
        
        // 모든 방문 후, 방문하지 않은 방이 있다면 false
        for (int i = 0; i < n; i++)
            if (!visited[i]) return false;
        return true;
    }
    
    private void dfs() {
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        visited[0] = true;

        for (int next : adjList.get(0)) stack.push(next);
        
        while(!stack.empty()) {
            int now = stack.pop();
            if (visited[now]) continue;
            
            if (!visited[before[now]]) {
                after[before[now]] = now;
                continue;
            }
            
            visited[now] = true;
            
            for (int next : adjList.get(now))
                if (!visited[next]) stack.push(next);

            stack.push(after[now]);
        }
    }
}